File: num-comb/comb.rb
##############################
# comb.rb
def newGap gap
    gap /= 1.3;
    gap = gap.to_i;
    return 11 if gap == 9 or gap == 10
    return 1 if gap < 1
    gap
end
def combSort a
    len = a.length
    gap = len
    swapped = false
    begin
        swapped = false
        gap = newGap gap
        (0...(len-gap)).each do |i|
            if a[i] > a[i+gap]
                swapped = true
                a[i], a[i+gap] = a[i+gap], a[i]
            end
        end
    end while gap > 1 or swapped
end
N = 10000000;
arr = (1..N).to_a.reverse!
combSort arr
(1...N).each do |z|
    print '!' if arr[z]<arr[z-1]
end

File: num-comb/comb.lua
------------------------------
-- comb.lua
newGap = function(gap)
    gap = math.floor(gap / 1.3)
    if gap == 9 or gap == 10 then return 11 end
    if gap < 1 then return 1 end
    return gap
end
combSort = function(a, len)
    gap = len
    repeat
        swapped = false
        gap = newGap(gap)
        for i = 0, (len-gap-1) do
            j = i+gap
            if a[i] > a[j] then
                swapped = true                
                a[i], a[j] = a[j], a[i]
            end
        end
    until gap <= 1 and not swapped
end
N = 10000000
arr = {}
for z=0, N-1 do arr[z] = N-z end
combSort(arr,N)
for z=1, N-1 do 
    if arr[z]<arr[z-1] then print("!") end 
end
